atmo/provisioners.py

```diff
@@ -0,0 +1,108 @@
+        self.config = settings.AWS_CONFIG
+        self.spark_emr_configuration_url = (
+            'https://s3-%s.amazonaws.com/%s/configuration/configuration.json' %
+            (self.config['AWS_REGION'], self.config['SPARK_EMR_BUCKET'])
+        )
+        self.script_uri = (
+            's3://%s/bootstrap/telemetry.sh' % self.config['SPARK_EMR_BUCKET']
+        )
+        self.emr = boto3.client(
+            'emr',
+            region_name=self.config['AWS_REGION'],
+        )
+        self.s3 = boto3.client(
+            's3',
+            region_name=self.config['AWS_REGION'],
+        )
+        self.session = requests.session()
+
+    def spark_emr_configuration(self):
+        """
+        Fetch the Spark EMR configuration data to be passed as the
+        Configurations parameter to EMR API endpoints.
+
+        We store this in S3 to be able to share it between various
+        Telemetry services.
+        """
+        response = self.session.get(self.spark_emr_configuration_url)
+        response.raise_for_status()
+        return response.json()
+
+    def job_flow_params(self, user_email, identifier, emr_release, size):
+        # setup instance groups using spot market for slaves
+        instance_groups = [
+            {
+                'Name': 'Master',
+                'Market': 'ON_DEMAND',
+                'InstanceRole': 'MASTER',
+                'InstanceType': self.config['MASTER_INSTANCE_TYPE'],
+                'InstanceCount': 1
+            }
+        ]
+
+        if size > 1:
+            core_group = {
+                'Name': 'Worker Instances',
+                'InstanceRole': 'CORE',
+                'InstanceType': self.config['WORKER_INSTANCE_TYPE'],
+                'InstanceCount': size,
+            }
+            if constance.config.AWS_USE_SPOT_INSTANCES:
+                core_group.update({
+                    'Market': 'SPOT',
+                    'BidPrice': str(constance.config.AWS_SPOT_BID_CORE),
+                })
+            else:
+                core_group['Market'] = 'ON_DEMAND'
+
+            instance_groups.append(core_group)
+
+        now = timezone.now().isoformat()
+
+        log_uri = (
+            's3://%s/%s/%s/%s' %
+            (self.config['LOG_BUCKET'], self.log_dir, identifier, now)
+        )
+
+        return {
+            'Name': str(uuid4()),
+            'LogUri': log_uri,
+            'ReleaseLabel': 'emr-%s' % emr_release,
+            'Configurations': self.spark_emr_configuration(),
+            'Instances': {
+                'InstanceGroups': instance_groups,
+                'Ec2KeyName': self.config['EC2_KEY_NAME'],
+                'KeepJobFlowAliveWhenNoSteps': True,  # same as no-auto-terminate
+            },
+            # TODO: this was missing in the job scheduling, but existing in the
+            # cluster provisioning paramters, should this be left out of jobs?
+            'JobFlowRole': self.config['SPARK_INSTANCE_PROFILE'],
```

Comment:
JobFlowRole? I see it's passed to run_job_flow in both spark_job_run and cluster_start on master. My understanding is that it defines the IAM role (aka aws permissions) for each node in the cluster.